src/Database/RocksDB/Base.hs

```diff
@@ -154,50 +169,75 @@ open path opts = liftIO $ bracketOnError initialize finalize mkDB
             when (createIfMissing opts) $
                 GHC.setFileSystemEncoding GHC.utf8
             pure (opts', oldenc)
-        finalize (opts', oldenc) = do
+        finalizeOpts (opts', oldenc) = do
             freeOpts opts'
             GHC.setFileSystemEncoding oldenc
 # endif
         mkDB (opts'@(Options' opts_ptr _ _), _) = do
             when (createIfMissing opts) $
                 createDirectoryIfMissing True path
-            withFilePath path $ \path_ptr ->
-                liftM (`DB` opts')
-                $ throwIfErr "open"
-                $ c_rocksdb_open opts_ptr path_ptr
+            withFilePath path $ \path_ptr -> do
+                db_ptr <- throwIfErr "open" $ c_rocksdb_open opts_ptr path_ptr
+                statusRef <- newIORef Open
+                return $ DB
+                    { db_ptr
+                    , dbOptions = opts'
+                    , dbHandleStatusRef = statusRef
+                    }
+
+-- | The simplest and safest way to use a DB.
+--
+-- The DB is closed when this function returns.
+--
+-- You must not return the DB out of the inner scope.
+--
+-- You must not call `close` on the DB.
+withDB :: (HasCallStack) => FilePath -> Options -> (DB -> IO a) -> IO a
+withDB path opts f = do
+    bracket
+        (open path opts)
+        close
+        f
 
 -- | Close a database.
 --
 -- The handle will be invalid after calling this action and should no
 -- longer be used.
-close :: MonadIO m => DB -> m ()
-close (DB db_ptr opts_ptr) = liftIO $
-    c_rocksdb_close db_ptr `finally` freeOpts opts_ptr
-
+-- If it is used, with the current implementation of RocksDB,
+-- the program will SEGFAULT.
+-- If another RocksDB operation is still running in another thread while
+-- close() is function is called, this may results in crashes or data loss.
+close :: (MonadIO m, HasCallStack) => DB -> m ()
+close db@DB{ db_ptr, dbOptions } = liftIO $ do
+    ensureOpenAndClose db
+    c_rocksdb_close db_ptr `finally` freeOpts dbOptions
```

Comment:
What if there is an async exception after `ensureOpenAndClose` and before this line? AFAIU, we will set status to `Closed`, but won't actually close the DB.
But on the other hand this function is intended to be used as a release action in `bracket` or anything similar, so it shouldn't be a problem I suppose.

Reply:
@gromakovsky That is correct.

None of the "plain" functions in this module are async-exception-safe; the same comment [as I wrote on `open`](https://github.com/serokell/rocksdb-haskell/pull/10/files/fc69de928e8fafb529e571320f88f3f6f4be1978#diff-56244378bea7e31bbe12cb39da7b646cR145) applies. It is not clear to me if the library authors had it this way by design (it might be that they planned for all plain functions to be intended to only be used in call-sites that themselves use `bracket`) or by accident, so I haven't made `close` use `mask_ $ do` so far.

I should probably copy-paste that comment here, though it might possibly have to be copied to a lot more functions (but I haven't analysed in detail which), so until now I have instead Haddocked the functions that _are_ async-exception-safe.

Only `withDB` and `openBracket` are `async-exception-safe`.

Reply:
What about changing 

```haskell
data DB = DB RocksDBPtr Options'
```

to

```haskell
data DB = DB (IORef (Maybe (RocksDBPtr, Options'))) 
```

And the `close`  can be implemented as:

```haskell
close (DB m_ref) = liftIO $ do
    atomicModifyIORef'
        m_ref
        (\m ->
             fmap
                 (Nothing, )
                 (case m of
                      Nothing -> pure ()
                      Just (db_ptr, opts_ptr) -> do
                          c_rocksdb_close db_ptr 
                          freeOpts opts_ptr))
```

This way:

1. You can't double close: in a single thread (or multiple threads; IORef's `atomicModifyIORef'` guarantees ordering).
2. Any other operation that needs a db_ptr (read/put/modify) can simply throw an exception if the m_ref contains a `Nothing`, complaining that the database has closed.
3. IORef operations are "guaranteed not to be interruptible", so you can't get an async exception while performing them.